src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -657,8 +673,134 @@ public override Task WriteAsync(Byte[] array, int offset, int count, Cancellatio
+
+                    Array.Clear(_arrayPoolBuffer, 0, _arrayPoolBufferHighWaterMark); // clear only the most we used
+                    ArrayPool<byte>.Shared.Return(_arrayPoolBuffer, clearArray: false);
+                }
+            }
+
+            public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            {
+                // Validate inputs
+                Debug.Assert(buffer != _arrayPoolBuffer);
+                _deflateStream.EnsureNotDisposed();
+                if (count <= 0)
+                {
+                    return;
+                }
+                else if (count > buffer.Length - offset)
+                {
+                    // The source stream is either malicious or poorly implemented and returned a number of
+                    // bytes larger than the buffer supplied to it.
+                    throw new InvalidDataException(SR.GenericInvalidData);
+                }
+
+                // Feed the data from base stream into the decompression engine.
+                _deflateStream._inflater.SetInput(buffer, offset, count);
+
+                // While there's more decompressed data available, forward it to the destination stream.
+                while (true)
+                {
+                    int bytesRead = _deflateStream._inflater.Inflate(_arrayPoolBuffer, 0, _arrayPoolBuffer.Length);
+                    if (bytesRead > 0)
+                    {
+                        if (bytesRead > _arrayPoolBufferHighWaterMark) _arrayPoolBufferHighWaterMark = bytesRead;
+                        await _destination.WriteAsync(_arrayPoolBuffer, 0, bytesRead, cancellationToken).ConfigureAwait(false);
+                    }
+                    else break;
+                }
+            }
+
+            public override void Write(byte[] buffer, int offset, int count) => WriteAsync(buffer, offset, count, default(CancellationToken)).GetAwaiter().GetResult();
+            public override bool CanWrite => true;
+            public override void Flush() { }
+
+            public override bool CanRead => false;
+            public override bool CanSeek => false;
+            public override long Length { get { throw new NotSupportedException(); } }
+            public override long Position { get { throw new NotSupportedException(); } set { throw new NotSupportedException(); } }
+            public override int Read(byte[] buffer, int offset, int count) { throw new NotSupportedException(); }
+            public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+            public override void SetLength(long value) { throw new NotSupportedException(); }
+        }
+
+        private void IncrementAsyncOperations()
+        {
+            int newCount = Interlocked.Increment(ref _asyncOperations);
+            Debug.Assert(newCount == 1, $"DeflateStream was misused, with multiple pending async operations");
```

Comment:
I wonder that as well, but I didn't want to mess with that as part of this change; I just encapsulated the check into its own method so that we can more easily do such things in the future (it also helps workaround a compiler warning regarding passing a ref on a MarshalByRefObject).  There are earlier checks for whether _asyncOperations != 0, throwing an exception in such cases... I don't know why those checks aren't just based on these atomic operations so as to avoid race conditions, but such a change now would be a small breaking change, as it could impact which exceptions get thrown in a very niche corner cases.  Probably not something we should care too much about, but definitely something to be left for later.